File: multithreading.jl
using Base.Threads
using Printf

const N = 1_000_000
const NUM_THREADS = 4

# Thread-safe array and lock
const primes = Vector{Int}()
const lock = ReentrantLock()

# Prime checking function
function is_prime(n::Int)
    if n ≤ 1
        return false
    elseif n == 2
        return true
    elseif iseven(n)
        return false
    end
    limit = isqrt(n)
    for i in 3:2:limit
        if n % i == 0
            return false
        end
    end
    return true
end

# Worker function (used within threads)
function find_primes_range(start::Int, stop::Int)
    local_primes = Int[]
    for n in start:stop
        if is_prime(n)
            push!(local_primes, n)
        end
    end
    lock(lock) do
        append!(primes, local_primes)
    end
end

# Main execution
function main()
    println(" Finding primes up to $N using $NUM_THREADS threads...")
    GC.gc()
    start_time = time()

    # Thread pool (manual thread splitting)
    chunk = div(N, NUM_THREADS)
    threads = []

    allocated = @allocated begin
        for i in 1:NUM_THREADS
            start_idx = (i - 1) * chunk + 1
            end_idx = (i == NUM_THREADS) ? N : i * chunk
            t = @spawn find_primes_range(start_idx, end_idx)
            push!(threads, t)
        end

        # Wait for all threads
        for t in threads
            fetch(t)
        end
    end

    elapsed = time() - start_time

    println(" Total primes found: ", length(primes))
    @printf("  Time taken       : %.2f seconds\n", elapsed)
    @printf(" Memory used      : %.2f MB\n", allocated / 1_000_000)
end

main()

File: multithreading.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <windows.h>
#include <psapi.h>

#define N 1000000
#define NUM_THREADS 4

int *primes;
int prime_count = 0;
CRITICAL_SECTION lock;

typedef struct {
    int start;
    int end;
} ThreadArgs;

int is_prime(int n) {
    if (n <= 1) return 0;
    if (n == 2) return 1;
    if (n % 2 == 0) return 0;
    int sqrt_n = (int)sqrt(n);
    for (int i = 3; i <= sqrt_n; i += 2) {
        if (n % i == 0) return 0;
    }
    return 1;
}

DWORD WINAPI find_primes(LPVOID param) {
    ThreadArgs* range = (ThreadArgs*)param;
    int start = range->start;
    int end = range->end;

    for (int i = start; i < end; ++i) {
        if (is_prime(i)) {
            EnterCriticalSection(&lock);
            primes[prime_count++] = i;
            LeaveCriticalSection(&lock);
        }
    }
    free(range);
    return 0;
}

double get_time_in_seconds() {
    FILETIME ft;
    GetSystemTimeAsFileTime(&ft);
    ULARGE_INTEGER uli;
    uli.LowPart = ft.dwLowDateTime;
    uli.HighPart = ft.dwHighDateTime;
    return uli.QuadPart / 10000000.0; // Convert to seconds
}

double get_memory_in_mb() {
    PROCESS_MEMORY_COUNTERS pmc;
    if (GetProcessMemoryInfo(GetCurrentProcess(), &pmc, sizeof(pmc))) {
        return pmc.PeakWorkingSetSize / (1024.0 * 1024.0);
    }
    return 0.0;
}

int main() {
    HANDLE threads[NUM_THREADS];
    primes = (int*)malloc(N * sizeof(int));
    if (!primes) {
        printf("Memory allocation failed.\n");
        return 1;
    }

    InitializeCriticalSection(&lock);

    double start_time = get_time_in_seconds();

    int chunk = N / NUM_THREADS;
    for (int i = 0; i < NUM_THREADS; ++i) {
        ThreadArgs* args = malloc(sizeof(ThreadArgs));
        args->start = i * chunk;
        args->end = (i == NUM_THREADS - 1) ? N : (i + 1) * chunk;
        threads[i] = CreateThread(NULL, 0, find_primes, args, 0, NULL);
    }

    WaitForMultipleObjects(NUM_THREADS, threads, TRUE, INFINITE);

    double end_time = get_time_in_seconds();
    double elapsed = end_time - start_time;
    double memory_used = get_memory_in_mb();

    printf("\nTotal primes found : %d\n", prime_count);
    printf("Time taken          : %.4f seconds\n", elapsed);
    printf("Peak memory used   : %.2f MB\n", memory_used);

    DeleteCriticalSection(&lock);
    free(primes);

    return 0;
}


